app/controllers/subdomain_redirects_controller.rb

```diff
@@ -0,0 +1,17 @@
+class SubdomainRedirectsController < ApplicationController
+  before_action :subdomain_redirects, only:
+
+  def subdomain_redirects
+    state = params[:state]
+
+    if state.ends_with?(".gov.uk") || state.include?("127.0.0.1:3000") || state.include?("localhost:3000")
```

Comment:
Crap, I've just realised this is insecure: because it would permit a state of `https://maliciouswebsite.com/?something=gov.uk`. So we need to parse `state` as a URI, extract the domain part and check that. But the good news is, once we've done this we can more elegantly construct the redirect URI. Something like:

```ruby
uri = URI.parse(state)
if uri.hostname.ends_with?(".gov.uk") || request.host == uri.hostname
  uri.path = admin_google_oauth2_omniauth_callback_path
  uri.query = params.slice(:code, :scope, :authuser, :hd, :prompt).permit!.to_query
  redirect_to uri.to_s, allow_other_host: true
else
  ...
```